### DISCUSSION TITLE
K-Fold Cross Validation - Ultralytics YOLOv8 Docs

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#4079] [Docs]
# K-Fold Cross Validation - Ultralytics YOLOv8 Docs

An in-depth guide demonstrating the implementation of K-Fold Cross Validation with the Ultralytics ecosystem for object detection datasets, leveraging Python, YOLO, and sklearn.

https://docs.ultralytics.com/guides/kfold-cross-validation/

**Comment by akashAD98** [comment]
not able to tain the model,i was able to split data ,but the images are not movesd to repective folds,they are empty.
it would be great if you share the collab link with it

**Reply by ignacioaj** [reply]
Make sure all images are in the folder 'dataset_path\images' and all labels are in the folder 'dataset_path\labels', i.e, move all images from 'train\images', 'valid\images' and 'test\images' to a unique folder 'images' (same for the labels). Otherwise, it won't work.

**Reply by pderrenger** [reply]
@ignacioaj it sounds like there might be an issue with the directory structure or the data preparation step before initiating the training process. Ensure that your dataset is structured correctly, with all images placed in the `dataset_path/images` folder and all labels in the `dataset_path/labels` folder. This structure is necessary for the k-fold cross-validation to properly access and distribute the data into folds.

If your images and labels are correctly placed but the issue persists, double-check the script or commands you're using to initiate the k-fold process. There might be a parameter that's not set correctly or a step that's being overlooked.

For a more detailed guide and troubleshooting, please refer to the documentation on K-Fold Cross Validation in the Ultralytics YOLOv8 Docs. Unfortunately, I cannot provide a Colab link, but the documentation should give you a comprehensive understanding of how to set up and run k-fold cross-validation with YOLOv8. If you continue to encounter issues, consider opening an issue on the GitHub repository with detailed information about the steps you've taken and the problem you're facing, so the community can assist you further. 📘🛠️

**Comment by nadinator** [comment]
I believe there's a mistake in the very last step:
```
results = model.train(data=dataset_yaml, *args, **kwargs)  # Include any training arguments
```

I think this should just be:
```
model.train(data=dataset_yaml, *args, **kwargs)  # Include any training arguments
```

It would be a good idea to include a link to the github page for these docs so I can submit a PR directly, rather than having to go search for it.

**Reply by pdlje82** [reply]
True,

also
```
with open(yaml_file, 'r', encoding="utf8") as y:
    classes = yaml.safe_load(y)['names']
cls_idx = sorted(classes.keys())
```
should be somethig like:
```
with open(yaml_path, 'r', encoding="utf8") as y:
    classes = yaml.safe_load(y)['names']
cls_idx = list(range(len(classes))) 
```

**Reply by Burhan-Q** [reply]
@nadinator the final two lines are

```py
    model.train(data=dataset_yaml,epochs=epochs, batch=batch, project=project)  # include any train arguments
    results[k] = model.metrics  # save output metrics for further analysis
```
The idea is that the metrics are stored in a dictionary for retrieval later. That said, there are definitely other ways to accomplish this and this probably could be updated. If you were looking for the Docs Source, you can find them at [in this repo](https://github.com/ultralytics/ultralytics) under the `docs` directory.

**Reply by Burhan-Q** [reply]
@pdlje82 if you're referring to the third step, the YOLO dataset YAML format should have the keys as integers and the values as the class-names, which is why it's done the way it is. For instance, if the classes were indexed at intervals of ten 

```yaml
names:
  10: one
  20: two
  30: three
  ...
```
the method you outline would not be able to correctly match the class index number to the proper class label. That said, I imagine it's unlikely that this labeling schema would be common, but it's not impossible to rule out either.

**Comment by pdlje82** [comment]
How can this be done System Memory efficient (GoogleColab)? After each fold run more System RAM is used and even with 24GB RAM, after 3 folds the RAM is full and the kernel crashes. GC doesnt seem to work and not even %reset -f did free the RAM.

**Reply by nadinator** [reply]
This might help: https://github.com/ultralytics/yolov5/issues/8397

**Reply by pderrenger** [reply]
@nadinator when running K-Fold Cross Validation on a platform like Google Colab, it's important to manage memory efficiently to prevent crashes. Here are some tips to help you with memory management:

1. **Garbage Collection**: After each fold, explicitly call Python's garbage collector using `import gc; gc.collect()`. This can help to free up unreferenced memory.

2. **Del Variables**: Delete large variables that are no longer needed with `del variable_name` and then call the garbage collector.

3. **Model Release**: If you're creating a new model instance for each fold, ensure you release the model and its resources by setting it to `None` and then calling the garbage collector.

4. **Data Loader**: When using data loaders, make sure to properly shut them down after each fold to free up memory.

5. **Torch's Cache**: Clear PyTorch's cache with `torch.cuda.empty_cache()` if you're using a GPU.

6. **Reduce Batch Size**: A smaller batch size will use less memory, though it may affect training stability and performance.

7. **Checkpointing**: Save model checkpoints after each fold to disk and then clear the memory. This way, if a crash occurs, you can resume without starting over.

8. **Lightweight Validation**: During validation, use a smaller subset of the data if possible to reduce memory usage.

9. **Streamline Code**: Ensure your code doesn't have memory leaks by avoiding unnecessary data copies and consolidating operations.

10. **Restart Runtime**: As a last resort, you can restart the Colab runtime between folds to clear all memory, but you'll need to save your progress externally.

Remember that while these strategies can help, there are inherent limitations to the amount of memory available in environments like Google Colab. If you consistently run out of memory, you may need to consider using a platform with more resources.

**Comment by kikazaki** [comment]
I think there is a mistake in the training loop in the end. k-fold cross-validation should train a new model from scratch k times to compare/average results. But in this loop, the same model is being trained on iteratively k times (no resetting of weights after each training iteration).

I think this line should be modified. However I am not familiar enough with the library to suggest a solution?
`model.train(data=dataset_yaml,epochs=epochs, batch=batch, project=project)  # include any train arguments`

**Reply by Burhan-Q** [reply]
@kikazaki you're correct, the model weights need to be either reset or overwritten with a fixed pretrained weights file. I'll put this addition on my to-do list, thank you for bringing it up!

**Comment by uri-el99** [comment]
I've got a doubt. When I'm trying to use K-Fold, in the step
`with open(yaml_file, 'r', encoding="utf8") as y:
    classes = yaml.safe_load(y)['names']
cls_idx = sorted(classes.keys())`
I got a error related to `AttributeError: 'list' object has no attribute 'keys' in cls_idx = sorted(classes.keys())`
It is worth mentioning that I am working with a single class.
How should I place my class? Hope you can help me out.

**Reply by Burhan-Q** [reply]
@uri-el99 see [my response here](https://github.com/orgs/ultralytics/discussions/4079#discussioncomment-7708444) on how to handle instances where the `names` are a list and not a dictionary as expected. The dictionary format is what is specified in the [Detection Datasets](https://docs.ultralytics.com/datasets/detect/) section of the official Documentation, so this guide is written following the official format.

**Comment by DP1701** [comment]
Hi @glenn-jocher and @Burhan-Q,

Thanks for the tutorial. I might have two suggestions for improvement: 

First suggestion: 

```python
yaml_file = 'path/to/data.yaml'  # your data YAML with data directories and names dictionary
with open(yaml_file, 'r', encoding="utf8") as y:
    classes = yaml.safe_load(y)['names']
cls_idx = sorted(classes.keys())
```

to

```python
with open(path_to_yaml, 'r', encoding="utf8") as y:
    classes = yaml.safe_load(y)['names']
classes = {index: value for index, value in enumerate(classes)} # Convert list to dict (keys contain indices)
cls_idx = classes.keys()
```

The problem is actually that ```classes``` is a list at the beginning but ```cls_index``` requires the keys of a dictionary. At least the script works for me after the change. 


Second suggestion:


```python
for ext in supported_extensions:
    images.extend(sorted((dataset_path / 'images').rglob(f"*{ext}")))
```

to

```python
for ext in supported_extensions:
    images.extend(sorted((dataset_path).rglob(f"*{ext}")))
```

In this way, not all images have to be in a single folder with the name "image". In the case of the Fruit_Detection dataset, this works. 
What do you think about these changes? Do you see other problems as a result?

My question now is what is the best way to get a result later, to get information about the performance of the trained model. With the script there are 5 separate results.

**Reply by Burhan-Q** [reply]
@DP1701 thanks for your input. I just so happened to be on my GH notifications page when you added your comment 😂

I certainly think that I should include some updates changes from the original implementation of the guide. That said, here are my thoughts on your suggestions.

```diff
with open(path_to_yaml, 'r', encoding="utf8") as y:
    classes = yaml.safe_load(y)['names']
- classes = {index: value for index, value in enumerate(classes)} # Convert list to dict (keys contain indices)
+ classes = classes if isinstance(classes, dict) else {index: value for index, value in enumerate(classes)}
cls_idx = classes.keys()
```
The original assumption was that the YAML would already be in the key-value dictionary format using indices for keys. Given that there might be occasions where that might not be the case, your point is valid however if `names` _is already_ a dictionary, you'd end up with a dictionary of indices key-values `{0:0, 1:1, ..., n:n}` since iterating a dictionary, the `value` would actually be a key. So adding a check for if `classes` is a dictionary will cover the original assumption and other common iterable type.

```py
(dataset_path / 'images').rglob(f"*{ext}")
```
Again, the assumption here is that all images would be found in `'Fruit-detection/images'` just like all the label files are expected to be in `'Fruit-detection/labels'` but the only assumption made about the filenames is that the label text file name would match the image file name. The reason I originally left out looking for all images in a dataset directory was to ensure that if there were other images in the subdirectory (for example, a dataset stats plot), were not also collected.

Hopefully those replies make sense and help you understand a little about what the thinking is/was behind those choices. Like I said, my aim is to update this guide soon as there are change you and others have brought up, as well as some things I've felt should be updated/fixed.

As to your final question:

> My question now is what is the best way to get a result later, to get information about the performance of the trained model. With the script there are 5 separate results.

Each run will generate a set of results files. The default location would be something like `runs/detect/train1...train5` (5 directories with increasing indices), but in the example the `project = 'kfold_demo'` argument means it would be something like `kfold_demo/detect/train1...train5`. Additionally, in the [Train YOLO using K-Fold Data Splits](https://docs.ultralytics.com/guides/kfold-cross-validation/#train-yolo-using-k-fold-data-splits) step (2) shows the `model.metrics` being added to the `results` dictionary. The dictionary could also be saved to disk and/or modified however you'd like. One *critical* point made by another commenter is that the model should get re-initialized with the pretrained weights before starting the next KFold-training.

```diff
for k in range(ksplit):
    dataset_yaml = ds_yamls[k]
+  model = YOLO(weights_path, task='detect') # load pretrained starting weights, load from scratch, or reset weights
    model.train(data=dataset_yaml,epochs=epochs, batch=batch, project=project)  # include any train arguments
    results[k] = model.metrics  # save output metrics for further analysis
```

**Reply by DP1701** [reply]
@Burhan-Q 
Thanks for the quick answer!

Ok, so I have k=5 in total 5 validation results. From these results (mAP, Precision, Recall, ...) the average should be calculated. Or am I misunderstanding this?

**Reply by Burhan-Q** [reply]
@DP1701 yes, if you're looking for an aggregate performance value of how the model performs on all the data, taking the average and standard deviation on these metrics would give you that value.

**Reply by DP1701** [reply]
@Burhan-Q Thank you for your answer!

One question: Do you have any idea how to implement the whole thing with StratifiedKFold? Changing ```kf = KFold(n_splits=ksplit, shuffle=True, random_state=20)``` to ```kf = StratifiedKFold(n_splits=ksplit, shuffle=True, random_state=20)``` is unfortunately not enough, since ```kfolds = list(kf.split(labels_df))``` requires input for both the ```X``` and ```y``` arguments. Sorry for this question, but somehow I can't get any further.

**Reply by Burhan-Q** [reply]
@DP1701 I do not unfortunately. I was never able to get `StratifiedKFold` to work with a detection dataset using this type of method in the experimenting I did. What I did instead was adjusted the `random_state` until the distribution of the data looked _close enough_ to the target I was aiming for and there were no gratuitous discrepancies in the class distribution.

**Comment by tataganesh** [comment]
Small suggestion: I believe the [Fruit Detection](https://www.kaggle.com/datasets/lakshaytyagi01/fruit-detection/code) dataset should point to the [Data Card ](https://www.kaggle.com/datasets/lakshaytyagi01/fruit-detection/data) instead of the code section.

**Reply by pderrenger** [reply]
@tataganesh thank you for your suggestion! 🙌 We always appreciate community feedback to improve our documentation and resources. I'll make sure to pass this on to our documentation team for review. If you have any more insights or require further assistance, feel free to reach out. Happy coding with YOLOv8! 🚀🍎

**Comment by tataganesh** [comment]
Thank you for the wonderful tutorial! I adapted the guide to avoid copying the images for different splits, and instead created different train and val text files for each split to perform cross-validation. Here is the corresponding [notebook](https://www.kaggle.com/code/tataganesh/k-fold-cross-validation-and-yolov8) detailing this approach!

**Reply by Burhan-Q** [reply]
@tataganesh thank you for sharing your notebook and for the compliments on this guide! I'll have to take a closer look as I've been thinking about doing a refresh on this guide for a while to incorporate changes discussed in the comments here and other ideas I've had. When I do my revision, I'll be certain to come back and reference your code as I agree that copy the images might not be desirable or feasible in all cases.

**Reply by uri-el99** [reply]
@Burhan-Q Thank you for all the support and for resolving doubts, it is a great job.
I would appreciate it if in future updates you could add a procedure not only for detection, but also for segmentation, that would be fantastic! Thank you so much.

**Comment by N786h** [comment]
After training, how do I get the generalized results and best.pt file?

**Reply by pderrenger** [reply]
@N786h after training your YOLOv8 model, the results of the training process, including metrics like precision, recall, and mAP (mean Average Precision), are automatically logged. The `best.pt` file, which contains the weights of the model that achieved the best validation performance during training, is saved in the `runs/train/exp*/weights` directory by default. You can access the generalized results by reviewing the `results.csv` file located in the same `runs/train/exp*` directory, which provides a detailed log of the training and validation metrics for each epoch.

To further analyze the performance of your model, you can use the `val` mode to evaluate the model on the validation dataset:

```python
from ultralytics import YOLO

# Load the best model
model = YOLO('runs/train/exp*/weights/best.pt')

# Evaluate the model's performance on the validation set
results = model.val()
```

This will give you a clear understanding of how well your model generalizes to new data. 🚀

**Comment by TonSarun** [comment]
I applied the k-fold function to the model, and the function finished data splitting. After that, I started the training command with all fold data. Then, I observed warnings during training and, ultimately, encountered errors in the training results. (show as below)
How can I check and resolve this issue?

"train: WARNING ⚠️ /content/Fold_Cross_val/Fold_1/train/images/PF_4-9-_jpg.rf.ef7f2395c3268c5515f6276d8ee3ae94.jpg: ignoring corrupt image/label: cannot identify image file '/content/Fold_Cross_val/Fold_1/train/images/PF_4-9-_jpg.rf.ef7f2395c3268c5515f6276d8ee3ae94.jpg'
train: WARNING ⚠️ No labels found in /content/Fold_Cross_val/Fold_1/train/labels.cache. See https://docs.ultralytics.com/yolov5/tutorials/train_custom_data
train: New cache created: /content/Fold_Cross_val/Fold_1/train/labels.cache
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
/content/yolov5/train.py in <module>
    646 if __name__ == '__main__':
    647     opt = parse_opt()
--> 648     main(opt)

3 frames
/content/yolov5/utils/dataloaders.py in __init__(self, path, img_size, batch_size, augment, hyp, rect, image_weights, cache_images, single_cls, stride, pad, min_items, prefix)
    500             if cache['msgs']:
    501                 LOGGER.info('\n'.join(cache['msgs']))  # display warnings
--> 502         assert nf > 0 or not augment, f'{prefix}No labels found in {cache_path}, can not start training. {HELP_URL}'
    503 
    504         # Read cache

AssertionError: train: No labels found in /content/Fold_Cross_val/Fold_1/train/labels.cache, can not start training. See https://docs.ultralytics.com/yolov5/tutorials/train_custom_data"

**Reply by pderrenger** [reply]
@TonSarun it appears that there are issues with your dataset images and labels which are causing warnings and errors during training. Here are the steps you can take to resolve these issues:

1. **Corrupt Image File**: The warning about the corrupt image file suggests that one or more of your images cannot be opened. You should verify that the image `/content/Fold_Cross_val/Fold_1/train/images/PF_4-9-_jpg.rf.ef7f2395c3268c5515f6276d8ee3ae94.jpg` is not corrupted and is in a readable format. You can do this by trying to open it with an image viewer or using an image processing library like PIL in Python.

2. **Missing Labels**: The warning `No labels found in /content/Fold_Cross_val/Fold_1/train/labels.cache` indicates that the training process is unable to find label files for your images. Ensure that your label files are correctly named, formatted, and located in the expected directory. The label files should be in the same directory as specified in your dataset YAML file and should have the same base name as their corresponding images.

3. **Cache File**: The creation of a new cache file suggests that the dataset has changed or the cache file is being built for the first time. If you have made changes to your dataset, such as adding or removing images or labels, you may need to clear the existing cache file so that it can be regenerated correctly.

4. **Assertion Error**: The assertion error at the end is a result of the previous warnings. The training cannot start because it has not found any labels for your images. After addressing the issues with your images and labels, this error should be resolved.

To troubleshoot, you can:

- Check the integrity of your images and labels.
- Verify that your dataset is structured correctly, with labels matching the format expected by YOLOv8.
- Ensure that the paths in your dataset YAML file are correct and point to the right directories.
- If necessary, delete the existing cache file to allow the creation of a new one.

If you continue to experience issues, you may want to refer to the documentation for additional guidance on preparing your dataset and troubleshooting common problems. Remember to check the dataset preparation guidelines carefully, as they are crucial for successful training. 📚🛠️
